### DISCUSSION TITLE
Understanding the impact of Stage 1 on performance

### DISCUSSION THREAD

**Discussion by zakcutner** [simdjson/simdjson#1924] [Q&A] [answered]
Thank you for creating this project, I find it really fascinating! I also enjoyed reading the associated paper, but this left me with a general question on the impact of Stage 1 on the performance of the parser.

The techniques used to complete Stage 1 are really cool, and I understand why they are great for performance. However, I'm not quite clear on why the output from Stage 1 (the indexes of the structural and pseudo-structural characters) actually makes Stage 2 faster than if it just iterated over the input bytes only, like a regular JSON parser. For example, in the following snippet:

```json
{ "foo": true }
```

Stage 1 would give us the indexes `0 ({), 2 ("), 7 (:), 9 (t), 14 (})`. Let's use the point in Stage 2 at which we have got to the `t` as an example. Once we have checked that the next three characters are `rue`, we can use the indexes we were given by Stage 1 to hop directly to the `}`. 

However, aside from skipping whitespace, I don't see how this provides an advantage—we had to scan the `rue` anyways in order to check the validity of `true` so surely the only byte we end up not needing to identify is the whitespace byte? Although, I noticed that the paper discusses how simdjson still provides an, albeit smaller, performance advantage even for minified JSON documents (where we would expect very little whitespace).

Perhaps my example is just too basic to see the advantage, but I would be really interested to learn what I'm missing. I realise my question is more about the paper than the implementation in this library, so apologies in advance if it's a bit off topic!

**Comment by lemire** [comment, marked as answer]
In your mental model, do you take into account that processors are superscalar? E.g., you can both check that a string starts with true, and check that a character is } during the same cycle (or, at least, the operations overlap).

My laptop can easily sustain eight instructions per cycle. So whenever you analyze a routine from a performance point of view, you should think about how to break data dependencies.

Does that help?

The simdjson model is not necessary to get great performance but taking into account superscalarity is primordial.

**Reply by zakcutner** [reply]
Thanks for the response, I hadn't considered superscalar execution. That makes a lot more sense!
